sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala

```diff
@@ -217,6 +217,18 @@ class CodegenContext {
     splitExpressions(expressions = initCodes, funcName = "init", arguments = Nil)
   }
 
+  /**
+   * Return true if a given variable has been described as a global variable
+   */
+  def isDeclaredMutableState(varName: String): Boolean = {
```

Comment:
let's only enable this check in test environment, in case it has bugs and break production jobs.

Reply:
Of course, this PR uses the method only in `assert`.

Reply:
hmm, even only in `assert`, it still can fail compilation, doesn't it?

Reply:
How about checking Utils.isTesting and throwing an exception in tests?

Reply:
I think that it is a good way at caller side since this function is valid at debug and production environments.